consensus/polybft/slashing/double_signing_tracker.go

```diff
@@ -0,0 +1,218 @@
+}
+
+type Messages struct {
+	content MessagesMap
+	mux     sync.RWMutex
+}
+
+// getSenderMsgsLocked returns commit messages for given height, round and sender
+func (m *Messages) getSenderMsgsLocked(view *ibftProto.View, sender types.Address) []*ibftProto.Message {
+	if roundMsgs, ok := m.content[view.Height]; ok {
+		if senderMsgsMap, ok := roundMsgs[view.Round]; ok {
+			return senderMsgsMap[sender]
+		}
+	}
+
+	return nil
+}
+
+// addMessage registers message for the given height, round and sender
+func (m *Messages) addMessage(view *ibftProto.View, sender types.Address, msg *ibftProto.Message) {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+
+	var senderMsgs SenderMessagesMap
+
+	if roundMsgs, ok := m.content[view.Height]; !ok {
+		roundMsgs = make(map[uint64]SenderMessagesMap)
+		m.content[view.Height] = roundMsgs
+		senderMsgs = createSenderMsgsMap(view.Round, roundMsgs)
+	} else if senderMsgs, ok = roundMsgs[view.Round]; !ok {
+		senderMsgs = createSenderMsgsMap(view.Round, roundMsgs)
+	}
+
+	msgs := m.getSenderMsgsLocked(msg.View, sender)
+	if msgs == nil {
+		msgs = []*ibftProto.Message{msg}
+	} else {
+		msgs = append(msgs, msg)
+	}
+
+	senderMsgs[sender] = msgs
+}
+
+// createSenderMsgsMap initializes senders message map for the given round
+func createSenderMsgsMap(round uint64, roundMsgs map[uint64]SenderMessagesMap) SenderMessagesMap {
+	sendersMsgs := make(SenderMessagesMap)
+	roundMsgs[round] = sendersMsgs
+
+	return sendersMsgs
+}
+
+type DoubleSigningTracker interface {
+	Handle(msg *ibftProto.Message)
+	GetEvidences(height uint64) []*DoubleSignEvidence
+	PruneMsgsUntil(height uint64)
+}
+
+type DoubleSigningTrackerImpl struct {
+	preprepare  *Messages
+	prepare     *Messages
+	commit      *Messages
+	roundChange *Messages
+
+	logger hclog.Logger
+}
+
+func NewDoubleSigningTracker(logger hclog.Logger) *DoubleSigningTrackerImpl {
+	return &DoubleSigningTrackerImpl{
+		logger:      logger,
+		preprepare:  &Messages{content: make(MessagesMap)},
+		prepare:     &Messages{content: make(MessagesMap)},
+		commit:      &Messages{content: make(MessagesMap)},
+		roundChange: &Messages{content: make(MessagesMap)},
+	}
+}
+
+// Handle is implementation of IBFTMessageHandler interface, which handles IBFT consensus messages
+func (t *DoubleSigningTrackerImpl) Handle(msg *ibftProto.Message) {
+	if err := t.validateMsg(msg); err != nil {
+		t.logger.Debug("[ERROR] invalid IBFT message retrieved", "error", err, "message", msg)
+
+		return
+	}
+
+	sender := types.BytesToAddress(msg.From)
+
+	msgMap := t.resolveMessagesStorage(msg.GetType())
+	msgMap.addMessage(msg.View, sender, msg)
+}
+
+// PruneMsgsUntil deletes all messages maps until the specified height
+func (t *DoubleSigningTrackerImpl) PruneMsgsUntil(height uint64) {
+	for _, msgType := range ibftProto.MessageType_value {
+		msgs := t.resolveMessagesStorage(ibftProto.MessageType(msgType))
+		if msgs == nil {
+			continue
+		}
+
+		for msgHeight := range msgs.content {
```

Comment:
Should we also acquire msgs.mux here?

Reply:
Fixed https://github.com/0xPolygon/polygon-edge/pull/1756/commits/f34d131a162e6fe45a6bcdbf5d0ebbf1b96ce58f